### PR TITLE
[flang][OpenMP] Allow loop iteration variables in DSA clauses

### LINKED ISSUE #78938: [Flang][OpenMP] Incorrect execution result of a do-variable defined as shared in parallel construct
```
Version of flang-new : 18.0.0(0fe86f9c518fb1296bba8d66ce495f9dfff2c435)

```

A `do-variable` defined as `shared` in `parallel` construct  has an incorrect value after executing `parallel` construct.  
A `do-variable` is executing by `private`, not `shared`.

The following are the test program, Flang-new, Gfortran and ifort compilation/execution result.

snggz567_22.f90:
```fortran
subroutine s1
  ip=100
  ip2=-100
!$omp parallel shared(ip) private(ip2)
  DO ip = 0, 1
  END DO
  DO ip2 = 0, 1
  END DO
!$omp end parallel
  print *,'ip  shared  :',ip
  print *,'ip2 private :',ip2
end subroutine s1

program main
  call s1
end program main
```

```
$ export OMP_NUM_THREADS=2; flang-new -fopenmp snggz567_22.f90; ./a.out
 ip  shared  : 100
 ip2 private : -100
$
```

```
$ export OMP_NUM_THREADS=2; gfortran -fopenmp snggz567_22.f90; ./a.out
 ip  shared  :           2
 ip2 private :        -100
$
```

```
$ export OMP_NUM_THREADS=2; ifort -qopenmp snggz567_22.f90; ./a.out
 ip  shared  :           2
 ip2 private :        -100
$
```

### PR DESCRIPTION
**Pull request by luporl**
Iteration variables of non-associated loops may be listed in DSA
clauses.

Fixes https://github.com/llvm/llvm-project/issues/78938

### CODE REVIEW

File: flang/lib/Semantics/resolve-directives.cpp [hunk 1]
```diff
@@ -1649,6 +1649,12 @@ void OmpAttributeVisitor::ResolveSeqLoopIndexInParallelOrTaskConstruct(
       break;
     }
   }
+  // If this symbol already has a data-sharing attribute then there is nothing
+  // to do here.
+  if (const Symbol *symbol = iv.symbol)
+    for (auto symMap : targetIt->objectWithDSA)
+      if (symMap.first->name() == symbol->name())
+        return;
```
Comment (kiranchandramohan) [inline flang/lib/Semantics/resolve-directives.cpp:1657]: Nit: Frontend style uses curly braces.

**Review by kiranchandramohan — APPROVED** [review]
LGTM. Please wait for @mjklemm

**Review by mjklemm — APPROVED** [review]
LGTM!

**Comment by luporl** [comment]
Thanks for the quick reviews!

**Pull request merged**